In [1]:
#by convention, we use these shorter two-letter names
import pysal as ps
import pandas as pd
import numpy as np

# Spatial Data Processing with PySAL & Pandas

PySAL has two simple ways to read in data. But, first, you need to get the path from where your notebook is running on your computer to the place the data is. For example, to find where the notebook is running:

In [2]:
!pwd

/home/serge/Dropbox/p/pysal/workshops/scipy16/gds_scipy16/content/part1


PySAL has a command that it uses to get the paths of its example datasets. Let's work with a commonly-used dataset first. 

In [7]:
ps.examples.available()

['arcgis',
 'street_net_pts',
 'Point',
 'taz',
 'Line',
 'calemp',
 'baltim',
 'south',
 'burkitt',
 'mexico',
 'sids2',
 'geodanet',
 'newHaven',
 'networks',
 'snow_maps',
 'us_income',
 '10740',
 'chicago',
 'juvenile',
 'stl',
 'desmith',
 'virginia',
 'book',
 'nat',
 'Polygon',
 'columbus',
 'wmat',
 'sacramento2']

In [8]:
ps.examples.explain('us_income')

{'description': 'Per-capita income for the lower 47 US states 1929-2010',
 'explanation': [' * us48.shp: shapefile ',
  ' * us48.dbf: dbf for shapefile',
  ' * us48.shx: index for shapefile',
  ' * usjoin.csv: attribute data (comma delimited file)'],
 'name': 'us_income'}

In [9]:
csv_path = ps.examples.get_path('usjoin.csv')

In [10]:
f = ps.open(csv_path)
f.header

['Name',
 'STATE_FIPS',
 '1929',
 '1930',
 '1931',
 '1932',
 '1933',
 '1934',
 '1935',
 '1936',
 '1937',
 '1938',
 '1939',
 '1940',
 '1941',
 '1942',
 '1943',
 '1944',
 '1945',
 '1946',
 '1947',
 '1948',
 '1949',
 '1950',
 '1951',
 '1952',
 '1953',
 '1954',
 '1955',
 '1956',
 '1957',
 '1958',
 '1959',
 '1960',
 '1961',
 '1962',
 '1963',
 '1964',
 '1965',
 '1966',
 '1967',
 '1968',
 '1969',
 '1970',
 '1971',
 '1972',
 '1973',
 '1974',
 '1975',
 '1976',
 '1977',
 '1978',
 '1979',
 '1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009']

In [11]:
y2009 = f.by_col('2009')

In [12]:
y2009

[32274,
 32077,
 31493,
 40902,
 40093,
 52736,
 40135,
 36565,
 33086,
 30987,
 40933,
 33174,
 35983,
 37036,
 31250,
 35151,
 35268,
 47159,
 49590,
 34280,
 40920,
 29318,
 35106,
 32699,
 37057,
 38009,
 41882,
 48123,
 32197,
 46844,
 33564,
 38672,
 35018,
 33708,
 35210,
 38827,
 41283,
 30835,
 36499,
 33512,
 35674,
 30107,
 36752,
 43211,
 40619,
 31843,
 35676,
 42504]

### Working with shapefiles

We can also work with local files outside the built-in examples.

To read in a shapefile, we will need the path to the file.

In [13]:
shp_path = '../data/texas.shp'
print(shp_path)

../data/texas.shp


Then, we open the file using the `ps.open` command:

In [14]:
f = ps.open(shp_path)

`f` is what we call a "file handle." That means that it only *points* to the data and provides ways to work with it. By itself, it does not read the whole dataset into memory. To see basic information about the file, we can use a few different methods. 

For instance, the header of the file, which contains most of the metadata about the file:

In [15]:
f.header

{'BBOX Mmax': 0.0,
 'BBOX Mmin': 0.0,
 'BBOX Xmax': -93.50721740722656,
 'BBOX Xmin': -106.6495132446289,
 'BBOX Ymax': 36.49387741088867,
 'BBOX Ymin': 25.845197677612305,
 'BBOX Zmax': 0.0,
 'BBOX Zmin': 0.0,
 'File Code': 9994,
 'File Length': 49902,
 'Shape Type': 5,
 'Unused0': 0,
 'Unused1': 0,
 'Unused2': 0,
 'Unused3': 0,
 'Unused4': 0,
 'Version': 1000}

To actually read in the shapes from memory, you can use the following commands:

In [16]:
f.by_row(14) #gets the 14th shape from the file

In [17]:
all_polygons = f.read() #reads in all polygons from memory

In [18]:
len(all_polygons)

254

So, all 254 polygons have been read in from file. These are stored in PySAL shape objects, which can be used by PySAL and can be converted to other Python shape objects. ]

They typically have a few methods. So, since we've read in polygonal data, we can get some properties about the polygons. Let's just have a look at the first polygon:

In [19]:
all_polygons

In [20]:
all_polygons[0].centroid #the centroid of the first polygon

(-100.27156110567945, 36.27508640938005)

In [21]:
all_polygons[0].area

0.23682222998468205

In [22]:
all_polygons[0].perimeter

1.9582821721538344

While in the Jupyter Notebook, you can examine what properties an object has by using the tab key.

In [23]:
polygon = all_polygons[0]

In [16]:
polygon. #press tab when the cursor is right after the dot

SyntaxError: invalid syntax (<ipython-input-16-aa03438a2fa8>, line 1)

### Working with Data Tables

In [24]:

dbf_path = "../data/texas.dbf"
print(dbf_path)

../data/texas.dbf


When you're working with tables of data, like a `csv` or `dbf`, you can extract your data in the following way. Let's open the dbf file we got the path for above.

In [25]:
f = ps.open(dbf_path)

Just like with the shapefile, we can examine the header of the dbf file

In [26]:
f.header

['NAME',
 'STATE_NAME',
 'STATE_FIPS',
 'CNTY_FIPS',
 'FIPS',
 'STFIPS',
 'COFIPS',
 'FIPSNO',
 'SOUTH',
 'HR60',
 'HR70',
 'HR80',
 'HR90',
 'HC60',
 'HC70',
 'HC80',
 'HC90',
 'PO60',
 'PO70',
 'PO80',
 'PO90',
 'RD60',
 'RD70',
 'RD80',
 'RD90',
 'PS60',
 'PS70',
 'PS80',
 'PS90',
 'UE60',
 'UE70',
 'UE80',
 'UE90',
 'DV60',
 'DV70',
 'DV80',
 'DV90',
 'MA60',
 'MA70',
 'MA80',
 'MA90',
 'POL60',
 'POL70',
 'POL80',
 'POL90',
 'DNL60',
 'DNL70',
 'DNL80',
 'DNL90',
 'MFIL59',
 'MFIL69',
 'MFIL79',
 'MFIL89',
 'FP59',
 'FP69',
 'FP79',
 'FP89',
 'BLK60',
 'BLK70',
 'BLK80',
 'BLK90',
 'GI59',
 'GI69',
 'GI79',
 'GI89',
 'FH60',
 'FH70',
 'FH80',
 'FH90']

So, the header is a list containing the names of all of the fields we can read. If we were interested in getting the `['NAME', 'STATE_NAME', 'HR90', 'HR80']` fields.

If we just wanted to grab the data of interest, `HR90`, we can use either `by_col` or `by_col_array`, depending on the format we want the resulting data in:

In [27]:
HR90 = f.by_col('HR90')
print(type(HR90).__name__, HR90[0:5])
HR90 = f.by_col_array('HR90')
print(type(HR90).__name__, HR90[0:5])

list [0.0, 0.0, 18.31166453, 0.0, 3.6517674554]
ndarray [[  0.        ]
 [  0.        ]
 [ 18.31166453]
 [  0.        ]
 [  3.65176746]]


As you can see, the `by_col` function returns a list of data, with no shape. It can only return one column at a time:

In [28]:
HRs = f.by_col('HR90', 'HR80')

TypeError: __call__() takes 2 positional arguments but 3 were given

This error message is called a "traceback," as you see in the top right, and it usually provides feedback on why the previous command did not execute correctly. Here, you see that one-too-many arguments was provided to `__call__`, which tells us we cannot pass as many arguments as we did to `by_col`.

If you want to read in many columns at once and store them to an array, use `by_col_array`:

In [29]:
HRs = f.by_col_array('HR90', 'HR80')

In [30]:
HRs

array([[  0.        ,   0.        ],
       [  0.        ,  10.50199538],
       [ 18.31166453,   5.10386362],
       [  0.        ,   0.        ],
       [  3.65176746,  10.4297038 ],
       [  0.        ,   0.        ],
       [  0.        ,  18.85369532],
       [  2.59514448,   6.33617194],
       [  0.        ,   0.        ],
       [  5.59753708,   6.0331825 ],
       [ 17.3692707 ,  14.53139626],
       [  5.06893755,   4.9960032 ],
       [  9.73560868,   8.84307335],
       [  5.66989851,   4.67049647],
       [  0.        ,   0.        ],
       [  4.08893052,   4.44077341],
       [ 16.49348507,   0.        ],
       [  0.        ,   7.17154332],
       [  0.        ,   8.1799591 ],
       [  3.48074279,   7.8746358 ],
       [  0.        ,   6.0397415 ],
       [  3.67511944,  12.63104711],
       [  8.19705726,  17.14148582],
       [  0.        ,   0.        ],
       [ 11.19883532,   4.79616307],
       [  8.53606487,   5.95876534],
       [  6.30954634,   5.23450586],
 

It is best to use `by_col_array` on data of a single type. That is, if you read in a lot of columns, some of them numbers and some of them strings, all columns will get converted to the same datatype:

In [31]:
allcolumns = f.by_col_array(['NAME', 'STATE_NAME', 'HR90', 'HR80'])

In [32]:
allcolumns

array([['Lipscomb', 'Texas', '0.0', '0.0'],
       ['Sherman', 'Texas', '0.0', '10.501995379'],
       ['Dallam', 'Texas', '18.31166453', '5.1038636248'],
       ..., 
       ['Hidalgo', 'Texas', '7.3003167816', '8.2383277607'],
       ['Willacy', 'Texas', '5.6481219994', '7.6212251119'],
       ['Cameron', 'Texas', '12.302014455', '11.761321464']], 
      dtype='<U13')

Note that the numerical columns, `HR90` & `HR80` are now considered strings, since they show up with the single tickmarks around them, like `'0.0'`.

These methods work similarly for `.csv` files as well

### Using Pandas with PySAL

A new functionality added to PySAL recently allows you to work with shapefile/dbf pairs using Pandas. This *optional* extension is only turned on if you have Pandas installed. The extension is the `ps.pdio` module:

In [33]:
ps.pdio

<module 'pysal.contrib.pdutilities' from '/home/serge/anaconda2/envs/scipy16/lib/python3.5/site-packages/pysal/contrib/pdutilities/__init__.py'>

To use it, you can read in shapefile/dbf pairs using the `ps.pdio.read_files` command. 

In [34]:
shp_path = ps.examples.get_path('NAT.shp')
data_table = ps.pdio.read_files(shp_path)

This reads in *the entire database table* and adds a column to the end, called `geometry`, that stores the geometries read in from the shapefile. 

Now, you can work with it like a standard pandas dataframe.

In [35]:
data_table.head()

NAME  STATE_NAME STATE_FIPS CNTY_FIPS   FIPS  STFIPS  COFIPS  \
0  Lake of the Woods   Minnesota         27       077  27077      27      77   
1              Ferry  Washington         53       019  53019      53      19   
2            Stevens  Washington         53       065  53065      53      65   
3           Okanogan  Washington         53       047  53047      53      47   
4       Pend Oreille  Washington         53       051  53051      53      51   

   FIPSNO  SOUTH      HR60                        ...                          \
0   27077      0  0.000000                        ...                           
1   53019      0  0.000000                        ...                           
2   53065      0  1.863863                        ...                           
3   53047      0  2.612330                        ...                           
4   53051      0  0.000000                        ...                           

      BLK90      GI59      GI69      GI79      GI89       FH60  FH70  \
0  0.024534  0.285235  0.372336  0.342104  0.336455  11.279621   5.4   
1  0.317712  0.256158  0.360665  0.361928  0.360640  10.053476   2.6   
2  0.210030  0.283999  0.394083  0.357566  0.369942   9.258437   5.6   
3  0.155922  0.258540  0.371218  0.381240  0.394519   9.039900   8.1   
4  0.134605  0.243263  0.365614  0.358706  0.387848   8.243930   4.1   

        FH80       FH90                                           geometry  
0   5.663881   9.515860  <pysal.cg.shapes.Polygon object at 0x7f85bbdba...  
1  10.079576  11.397059  <pysal.cg.shapes.Polygon object at 0x7f85bbdba...  
2   6.812127  10.352015  <pysal.cg.shapes.Polygon object at 0x7f85bbdba...  
3  10.084926  12.840340  <pysal.cg.shapes.Polygon object at 0x7f85bbdba...  
4   7.557643  10.313002  <pysal.cg.shapes.Polygon object at 0x7f85bbdba...  

[5 rows x 70 columns]

The `read_files` function only works on shapefile/dbf pairs. If you need to read in data using CSVs, use pandas directly:

In [36]:
usjoin = pd.read_csv(csv_path)
#usjoin = ps.pdio.read_files(csv_path) #will not work, not a shp/dbf pair

In [37]:
usjoin.head()

Name  STATE_FIPS  1929  1930  1931  1932  1933  1934  1935  1936  \
0     Alabama           1   323   267   224   162   166   211   217   251   
1     Arizona           4   600   520   429   321   308   362   416   462   
2    Arkansas           5   310   228   215   157   157   187   207   247   
3  California           6   991   887   749   580   546   603   660   771   
4    Colorado           8   634   578   471   354   353   368   444   542   

   ...     2000   2001   2002   2003   2004   2005   2006   2007   2008   2009  
0  ...    23471  24467  25161  26065  27665  29097  30634  31988  32819  32274  
1  ...    25578  26232  26469  27106  28753  30671  32552  33470  33445  32077  
2  ...    22257  23532  23929  25074  26465  27512  29041  31070  31800  31493  
3  ...    32275  32750  32900  33801  35663  37463  40169  41943  42377  40902  
4  ...    32949  34228  33963  34092  35543  37388  39662  41165  41719  40093  

[5 rows x 83 columns]

The nice thing about working with pandas dataframes is that they have very powerful baked-in support for relational-style queries. By this, I mean that it is very easy to find things like:

The number of counties in each state:

In [38]:
data_table.groupby("STATE_NAME").size()

STATE_NAME
Alabama                  67
Arizona                  14
Arkansas                 75
California               58
Colorado                 63
Connecticut               8
Delaware                  3
District of Columbia      1
Florida                  67
Georgia                 159
Idaho                    44
Illinois                102
Indiana                  92
Iowa                     99
Kansas                  105
Kentucky                120
Louisiana                64
Maine                    16
Maryland                 24
Massachusetts            12
Michigan                 83
Minnesota                87
Mississippi              82
Missouri                115
Montana                  55
Nebraska                 93
Nevada                   17
New Hampshire            10
New Jersey               21
New Mexico               32
New York                 58
North Carolina          100
North Dakota             53
Ohio                     88
Oklahoma                 77
Oregon   

Or, to get the rows of the table that are in Arizona, we can use the `query` function of the dataframe:

In [39]:
data_table.query('STATE_NAME == "Arizona"')

NAME STATE_NAME STATE_FIPS CNTY_FIPS   FIPS  STFIPS  COFIPS  \
1707      Navajo    Arizona         04       017  04017       4      17   
1708    Coconino    Arizona         04       005  04005       4       5   
1722      Mohave    Arizona         04       015  04015       4      15   
1726      Apache    Arizona         04       001  04001       4       1   
2002     Yavapai    Arizona         04       025  04025       4      25   
2182        Gila    Arizona         04       007  04007       4       7   
2262    Maricopa    Arizona         04       013  04013       4      13   
2311    Greenlee    Arizona         04       011  04011       4      11   
2326      Graham    Arizona         04       009  04009       4       9   
2353       Pinal    Arizona         04       021  04021       4      21   
2499        Pima    Arizona         04       019  04019       4      19   
2514     Cochise    Arizona         04       003  04003       4       3   
2615  Santa Cruz    Arizona         04       023  04023       4      23   
3080      La Paz    Arizona         04       012  04012       4      12   

      FIPSNO  SOUTH       HR60  \
1707    4017      0   5.263989   
1708    4005      0   3.185449   
1722    4015      0   0.000000   
1726    4001      0  10.951223   
2002    4025      0   3.458771   
2182    4007      0   6.473749   
2262    4013      0   6.179259   
2311    4011      0   2.896284   
2326    4009      0   4.746648   
2353    4021      0  13.828390   
2499    4019      0   5.520841   
2514    4003      0   4.845049   
2615    4023      0   9.252406   
3080    4012      0   5.046682   

                            ...                             BLK90      GI59  \
1707                        ...                          0.905251  0.366863   
1708                        ...                          1.469081  0.301222   
1722                        ...                          0.324075  0.279339   
1726                        ...                          0.162361  0.395913   
2002                        ...                          0.298011  0.289509   
2182                        ...                          0.246171  0.265294   
2262                        ...                          3.499221  0.277828   
2311                        ...                          0.349650  0.177691   
2326                        ...                          1.890487  0.310256   
2353                        ...                          3.134586  0.304294   
2499                        ...                          3.118252  0.268266   
2514                        ...                          5.201590  0.261208   
2615                        ...                          0.326863  0.327130   
3080                        ...                          2.628811  0.271556   

          GI69      GI79      GI89       FH60  FH70       FH80       FH90  \
1707  0.414135  0.401999  0.445299  13.146998  12.1  13.762783  18.033782   
1708  0.377785  0.381655  0.403188   9.475171   8.5  11.181563  15.267643   
1722  0.347150  0.375790  0.374383  11.508554   4.8   7.018268   9.214294   
1726  0.450552  0.431013  0.489132  15.014738  14.6  18.727548  22.933635   
2002  0.378195  0.376313  0.384089   9.930032   8.6   7.516372   9.483521   
2182  0.337519  0.353848  0.386976  10.470261   8.1   9.934237  11.706102   
2262  0.352374  0.366015  0.372756  10.642382   9.8  11.857260  14.404902   
2311  0.257158  0.283518  0.337256   9.806115   6.7   5.295110  10.453284   
2326  0.362926  0.383554  0.408379  11.979335  10.1  11.961367  16.129032   
2353  0.369974  0.361193  0.400130  10.822965   8.8  10.341699  15.304144   
2499  0.367218  0.375039  0.392144  11.381626  10.2  12.689768  16.163178   
2514  0.359500  0.359701  0.399208  10.197573   8.7   9.912732  13.733872   
2615  0.396807  0.393240  0.413795  19.007213  14.7  15.690913  18.272244   
3080  0.364110  0.372662  0.405743   9.216414   8.0   9.296093  12.379134   

                                 

Behind the scenes, this uses a fast vectorized library, `numexpr`, to essentially do the following. 

First, compare each row's `STATE_NAME` column to `'Arizona'` and return `True` if the row matches:

In [40]:
data_table.STATE_NAME == 'Arizona'

0       False
1       False
2       False
3       False
4       False
5       False
6       False
7       False
8       False
9       False
10      False
11      False
12      False
13      False
14      False
15      False
16      False
17      False
18      False
19      False
20      False
21      False
22      False
23      False
24      False
25      False
26      False
27      False
28      False
29      False
        ...  
3055    False
3056    False
3057    False
3058    False
3059    False
3060    False
3061    False
3062    False
3063    False
3064    False
3065    False
3066    False
3067    False
3068    False
3069    False
3070    False
3071    False
3072    False
3073    False
3074    False
3075    False
3076    False
3077    False
3078    False
3079    False
3080     True
3081    False
3082    False
3083    False
3084    False
Name: STATE_NAME, dtype: bool

Then, use that to filter out rows where the condition is true:

In [41]:
data_table[data_table.STATE_NAME == 'Arizona']

NAME STATE_NAME STATE_FIPS CNTY_FIPS   FIPS  STFIPS  COFIPS  \
1707      Navajo    Arizona         04       017  04017       4      17   
1708    Coconino    Arizona         04       005  04005       4       5   
1722      Mohave    Arizona         04       015  04015       4      15   
1726      Apache    Arizona         04       001  04001       4       1   
2002     Yavapai    Arizona         04       025  04025       4      25   
2182        Gila    Arizona         04       007  04007       4       7   
2262    Maricopa    Arizona         04       013  04013       4      13   
2311    Greenlee    Arizona         04       011  04011       4      11   
2326      Graham    Arizona         04       009  04009       4       9   
2353       Pinal    Arizona         04       021  04021       4      21   
2499        Pima    Arizona         04       019  04019       4      19   
2514     Cochise    Arizona         04       003  04003       4       3   
2615  Santa Cruz    Arizona         04       023  04023       4      23   
3080      La Paz    Arizona         04       012  04012       4      12   

      FIPSNO  SOUTH       HR60  \
1707    4017      0   5.263989   
1708    4005      0   3.185449   
1722    4015      0   0.000000   
1726    4001      0  10.951223   
2002    4025      0   3.458771   
2182    4007      0   6.473749   
2262    4013      0   6.179259   
2311    4011      0   2.896284   
2326    4009      0   4.746648   
2353    4021      0  13.828390   
2499    4019      0   5.520841   
2514    4003      0   4.845049   
2615    4023      0   9.252406   
3080    4012      0   5.046682   

                            ...                             BLK90      GI59  \
1707                        ...                          0.905251  0.366863   
1708                        ...                          1.469081  0.301222   
1722                        ...                          0.324075  0.279339   
1726                        ...                          0.162361  0.395913   
2002                        ...                          0.298011  0.289509   
2182                        ...                          0.246171  0.265294   
2262                        ...                          3.499221  0.277828   
2311                        ...                          0.349650  0.177691   
2326                        ...                          1.890487  0.310256   
2353                        ...                          3.134586  0.304294   
2499                        ...                          3.118252  0.268266   
2514                        ...                          5.201590  0.261208   
2615                        ...                          0.326863  0.327130   
3080                        ...                          2.628811  0.271556   

          GI69      GI79      GI89       FH60  FH70       FH80       FH90  \
1707  0.414135  0.401999  0.445299  13.146998  12.1  13.762783  18.033782   
1708  0.377785  0.381655  0.403188   9.475171   8.5  11.181563  15.267643   
1722  0.347150  0.375790  0.374383  11.508554   4.8   7.018268   9.214294   
1726  0.450552  0.431013  0.489132  15.014738  14.6  18.727548  22.933635   
2002  0.378195  0.376313  0.384089   9.930032   8.6   7.516372   9.483521   
2182  0.337519  0.353848  0.386976  10.470261   8.1   9.934237  11.706102   
2262  0.352374  0.366015  0.372756  10.642382   9.8  11.857260  14.404902   
2311  0.257158  0.283518  0.337256   9.806115   6.7   5.295110  10.453284   
2326  0.362926  0.383554  0.408379  11.979335  10.1  11.961367  16.129032   
2353  0.369974  0.361193  0.400130  10.822965   8.8  10.341699  15.304144   
2499  0.367218  0.375039  0.392144  11.381626  10.2  12.689768  16.163178   
2514  0.359500  0.359701  0.399208  10.197573   8.7   9.912732  13.733872   
2615  0.396807  0.393240  0.413795  19.007213  14.7  15.690913  18.272244   
3080  0.364110  0.372662  0.405743   9.216414   8.0   9.296093  12.379134   

                                 

We might need this behind the scenes knowledge when we want to chain together conditions, or when we need to do spatial queries. 

This is because spatial queries are somewhat more complex. Let's say, for example, we want all of the counties in the US to the West of `-121` longitude. We need a way to express that question. Ideally, we want something like:

```
SELECT
        *
FROM
        data_table
WHERE
        x_centroid < -121
```

So, let's refer to an arbitrary polygon in the the dataframe's geometry column as `poly`. The centroid of a PySAL polygon is stored as an `(X,Y)` pair, so the longitude is the first element of the pair, `poly.centroid[0]`. 

Then, applying this condition to each geometry, we get the same kind of filter we used above to grab only counties in Arizona:

In [42]:
data_table.geometry.apply(lambda x: x.centroid[0] < -121)

0       False
1       False
2       False
3       False
4       False
5       False
6       False
7       False
8       False
9       False
10      False
11      False
12      False
13      False
14      False
15      False
16      False
17      False
18      False
19      False
20      False
21      False
22      False
23      False
24      False
25      False
26      False
27       True
28      False
29      False
        ...  
3055    False
3056    False
3057    False
3058    False
3059    False
3060    False
3061    False
3062    False
3063    False
3064    False
3065    False
3066    False
3067    False
3068    False
3069    False
3070    False
3071    False
3072    False
3073    False
3074    False
3075    False
3076    False
3077    False
3078    False
3079    False
3080    False
3081    False
3082    False
3083    False
3084    False
Name: geometry, dtype: bool

If we use this as a filter on the table, we can get only the rows that match that condition, just like we did for the `STATE_NAME` query:

In [54]:
data_table[data_table.geometry.apply(lambda x: x.centroid[0] < -119)]

NAME  STATE_NAME STATE_FIPS CNTY_FIPS   FIPS  STFIPS  COFIPS  \
3            Okanogan  Washington         53       047  53047      53      47   
27            Whatcom  Washington         53       073  53073      53      73   
31             Skagit  Washington         53       057  53057      53      57   
42             Chelan  Washington         53       007  53007      53       7   
44            Clallam  Washington         53       009  53009      53       9   
47          Snohomish  Washington         53       061  53061      53      61   
48             Island  Washington         53       029  53029      53      29   
55            Douglas  Washington         53       017  53017      53      17   
57          Jefferson  Washington         53       031  53031      53      31   
68              Grant  Washington         53       025  53025      53      25   
71             Kitsap  Washington         53       035  53035      53      35   
80               King  Washington         53       033  53033      53      33   
85              Mason  Washington         53       045  53045      53      45   
89           Kittitas  Washington         53       037  53037      53      37   
92       Grays Harbor  Washington         53       027  53027      53      27   
107            Pierce  Washington         53       053  53053      53      53   
116          Thurston  Washington         53       067  53067      53      67   
119            Yakima  Washington         53       077  53077      53      77   
130           Pacific  Washington         53       049  53049      53      49   
131             Lewis  Washington         53       041  53041      53      41   
139            Benton  Washington         53       005  53005      53       5   
169          Skamania  Washington         53       059  53059      53      59   
170           Cowlitz  Washington         53       015  53015      53      15   
171         Wahkiakum  Washington         53       069  53069      53      69   
181           Clatsop      Oregon         41       007  41007      41       7   
183          Columbia      Oregon         41       009  41009      41       9   
193             Clark  Washington         53       011  53011      53      11   
194         Klickitat  Washington         53       039  53039      53      39   
209            Morrow      Oregon         41       049  41049      41      49   
213           Gilliam      Oregon         41       021  41021      41      21   
...               ...         ...        ...       ...    ...     ...     ...   
1225          Douglas      Nevada         32       005  32005      32       5   
1233        El Dorado  California         06       017  06017       6      17   
1262             Yolo  California         06       113  06113       6     113   
1263           Alpine  California         06       003  06003       6       3   
1281             Napa  California         06       055  06055       6      55   
1282           Sonoma  California         06       097  06097       6      97   
1301       Sacramento  California         06       067  06067       6      67   
1316           Amador  California         06       005  06005       6       5   
1367        Calaveras  California         06       009  06009       6       9   
1381         Tuolumne  California         06       109  06109       6     109   
1401            Marin  California         06       041  06041       6      41   
1404      San Joaquin  California         06       077  06077       6      77   
1447           Solano  California         06       095  06095       6      95   
1462     Contra Costa  California         06       013  06013       6      13   
1466       Stanislaus  California         06       099  06099       6      99   
1514          Alameda  California         06       001  06001       6       1   
1515         Mariposa  California         06       043  06043       6      43   
1530    San Francisco  California         06    

## Other types of spatial queries

Everybody knows the following statements are true:

1. If you head directly west from Reno, you will shortly enter California.
2. San Diego is in California.

But what does this tell us about the location of San Diego relative to Reno?

Or for that matter, how many counties in California are to the east of Reno?





In [149]:
geom = data_table.query('(NAME == "Washoe") & (STATE_NAME == "Nevada")').geometry

In [150]:
lon,lat = geom.values[0].centroid

In [153]:
lon

-119.6555030699793

In [156]:
cal_counties = data_table.query('(STATE_NAME=="California")')

In [157]:
cal_counties[cal_counties.geometry.apply(lambda x: x.centroid[0] > lon)]

NAME  STATE_NAME STATE_FIPS CNTY_FIPS   FIPS  STFIPS  COFIPS  \
1312            Mono  California         06       051  06051       6      51   
1591          Fresno  California         06       019  06019       6      19   
1620            Inyo  California         06       027  06027       6      27   
1765          Tulare  California         06       107  06107       6     107   
1956            Kern  California         06       029  06029       6      29   
1957  San Bernardino  California         06       071  06071       6      71   
2117         Ventura  California         06       111  06111       6     111   
2255       Riverside  California         06       065  06065       6      65   
2279          Orange  California         06       059  06059       6      59   
2344       San Diego  California         06       073  06073       6      73   
2351     Los Angeles  California         06       037  06037       6      37   
2358        Imperial  California         06       025  06025       6      25   

      FIPSNO  SOUTH       HR60  \
1312    6051      0  15.062509   
1591    6019      0   5.192037   
1620    6027      0   8.558713   
1765    6107      0   3.364944   
1956    6029      0   6.393044   
1957    6071      0   3.243373   
2117    6111      0   3.180374   
2255    6065      0   4.898903   
2279    6059      0   2.083555   
2344    6073      0   2.387842   
2351    6037      0   4.564947   
2358    6025      0   4.160599   

                            ...                              BLK90      GI59  \
1312                        ...                           0.431900  0.229888   
1591                        ...                           5.007266  0.286651   
1620                        ...                           0.432143  0.242293   
1765                        ...                           1.480503  0.303303   
1956                        ...                           5.544117  0.273083   
1957                        ...                           8.103188  0.259038   
2117                        ...                           2.336118  0.259643   
2255                        ...                           5.433210  0.284000   
2279                        ...                           1.770587  0.255864   
2344                        ...                           6.377301  0.269147   
2351                        ...                          11.203381  0.267207   
2358                        ...                           2.398836  0.286568   

          GI69      GI79      GI89       FH60  FH70       FH80       FH90  \
1312  0.327520  0.388414  0.366316   6.743421   6.6   8.187135  10.083699   
1591  0.379884  0.394981  0.412947  11.788963  11.8  13.998747  18.523541   
1620  0.334361  0.353784  0.378516   9.676365   8.1   8.480065  12.067279   
1765  0.380229  0.395059  0.410811  10.398757   9.5  11.957928  16.303423   
1956  0.364603  0.380617  0.395393  10.225236  10.4  12.037755  16.118827   
1957  0.350927  0.362389  0.372659   9.857519  10.0  12.999831  15.403925   
2117  0.325908  0.350430  0.329175  10.304761   8.9  11.047165  12.044930   
2255  0.376737  0.383226  0.368177   9.769959   9.2  11.145612  12.678340   
2279  0.316731  0.355117  0.330447   8.802545   9.1  12.405423  12.974648   
2344  0.352428  0.382815  0.368301  11.425041  11.5  14.384523  15.489702   
2351  0.353837  0.402031  0.392669  13.071850  13.4  17.767317  18.808224   
2358  0.374913  0.397430  0.426348  11.466417   9.7  12.590249  17.243741   

                                               geometry  
1312  <pysal.cg.shapes.Polygon object at 0x7f85bb8a4...  
1591  <pysal.cg.shapes.Polygon object at 0x7f85bb794...  
1620  <pysal.cg.shapes.Polygon object at 0x7f85bb7b8...  
1765  <pysal.cg.shapes.Polygon object at 0x7f85bb743...  
1956  <pysal.cg.shapes.Polygon object at 0x7f85bb60c...  
1957  <pysal.cg.shapes.Polygon object at 0x7f85bb60c...  
2117  <pysal.cg.shapes.Polygon object at 0x7f85bb59d...  
2255  <pysal.cg.shapes.Po

In [158]:
len(cal_counties)

58

This works on any type of spatial query. 

For instance, if we wanted to find all of the counties that are within a threshold distance from an observation's centroid, we can do it in the following way. 

But first, we need to handle distance calculations on the earth's surface.

In [159]:

from numpy import sin,cos,arctan2,sqrt,pi # import from numpy

def gcd(loc1, loc2, R=3961):
    """Great circle distance
    
    Parameters
    ----------
    
    loc1: tuple (long, lat in decimal degrees)
    
    loc2: tuple (long, lat in decimal degrees)
    
    R: Radius of the earth (3961 miles, 6367 km)
    
    Returns
    -------
    great circle distance between loc1 and loc2 in units of R
    
    
    Notes
    ------
    Does not take into account non-spheroidal shape of the Earth
    
    Source: http://andrew.hedges.name/experiments/haversine/
    
    >>> san_diego = -117.1611, 32.7157
    >>> austin = -97.7431, 30.2672
    >>> gcd(san_diego, austin)
    1155.4746441646942
    
    """
    lon1, lat1 = loc1
    lon2, lat2 = loc2
    lon1 = lon1 * pi / 180.0
    lon2 = lon2 * pi / 180.0
    lat1 = lat1 * pi / 180.0
    lat2 = lat2 * pi / 180.0
    dlat = lat2 - lat1
    dlon = lon2 - lon1
    a = (sin(dlat/2.))**2 + cos(lat1) * cos(lat2) * (sin(dlon/2.))**2 
    c = 2 * arctan2( sqrt(a), sqrt(1-a) )
    return R * c
    

In [160]:
san_diego = -117.1611, 32.7157
austin = -97.7431, 30.2672
gcd(san_diego, austin)

1155.4746441646942

In [162]:
gcd(san_diego, austin, 6367) #km

1857.3357887898528

In [165]:

# Find all the counties with centroids within 50 miles of Austin
def near_target_point(polygon, target=austin, threshold=50):
    return gcd(polygon.centroid, target) < threshold 

data_table[data_table.geometry.apply(near_target_point)]

NAME STATE_NAME STATE_FIPS CNTY_FIPS   FIPS  STFIPS  COFIPS  \
2698      Burnet      Texas         48       053  48053      48      53   
2716  Williamson      Texas         48       491  48491      48     491   
2742      Travis      Texas         48       453  48453      48     453   
2751         Lee      Texas         48       287  48287      48     287   
2754      Blanco      Texas         48       031  48031      48      31   
2762     Bastrop      Texas         48       021  48021      48      21   
2769        Hays      Texas         48       209  48209      48     209   
2795    Caldwell      Texas         48       055  48055      48      55   
2798       Comal      Texas         48       091  48091      48      91   
2808   Guadalupe      Texas         48       187  48187      48     187   

      FIPSNO  SOUTH      HR60  \
2698   48053      1  0.000000   
2716   48491      1  9.511852   
2742   48453      1  4.242561   
2751   48287      1  7.449622   
2754   48031      1  0.000000   
2762   48021      1  3.938946   
2769   48209      1  5.016555   
2795   48055      1  9.677544   
2798   48091      1  3.359538   
2808   48187      1  5.743759   

                            ...                              BLK90      GI59  \
2698                        ...                           1.186224  0.327508   
2716                        ...                           4.916482  0.363603   
2742                        ...                          10.959791  0.299292   
2751                        ...                          13.847829  0.376002   
2754                        ...                           0.937709  0.369814   
2762                        ...                          11.792071  0.370264   
2769                        ...                           3.383424  0.385061   
2795                        ...                          10.704001  0.380080   
2798                        ...                           0.854684  0.274182   
2808                        ...                           5.649500  0.340374   

          GI69      GI79      GI89       FH60  FH70       FH80       FH90  \
2698  0.449285  0.385079  0.405890  10.774142   6.5   7.115629  10.568742   
2716  0.379902  0.341976  0.345201  13.532731   9.0   7.582572  12.032589   
2742  0.372293  0.378953  0.388149  12.976379  10.9  14.459691  17.307113   
2751  0.433132  0.394000  0.394959  12.305699  10.1   8.875542  10.530896   
2754  0.436449  0.394609  0.394414   9.365854   6.0   8.074074   9.080119   
2762  0.419933  0.390927  0.380907  14.747191  12.5  10.559006  12.281387   
2769  0.424973  0.378992  0.372266  13.064187   9.4  10.003691  10.337188   
2795  0.412614  0.410802  0.413940  15.229616  10.5  12.894034  17.191502   
2798  0.359174  0.375810  0.380032  11.315107   7.6   8.693149   9.104427   
2808  0.388821  0.354559  0.378073  11.758921   9.6  10.418149  12.918448   

                                               geometry  
2698  <pysal.cg.shapes.Polygon object at 0x7f85bb363...  
2716  <pysal.cg.shapes.Polygon object at 0x7f85bb388...  
2742  <pysal.cg.shapes.Polygon object at 0x7f85bb32f...  
2751  <pysal.cg.shapes.Polygon object at 0x7f85bb32f...  
2754  <pysal.cg.shapes.Polygon object at 0x7f85bb32f...  
2762  <pysal.cg.shapes.Polygon object at 0x7f85bb32f...  
2769  <pysal.cg.shapes.Polygon object at 0x7f85bb2d7...  
2795  <pysal.cg.shapes.Polygon object at 0x7f85bb2d7...  
2798  <pysal.cg.shapes.Polygon object at 0x7f85bb2d7...  
2808  <pysal.cg.shapes.Polygon object at 0x7f85bb2ff...  

[10 rows x 70 columns]

### Moving in and out of the dataframe

Most things in PySAL will be explicit about what type their input should be. Most of the time, PySAL functions require either lists or arrays. This is why the file-handler methods are the default IO method in PySAL: the rest of the computational tools are built around their datatypes. 

However, it is very easy to get the correct datatype from Pandas using the `values` and `tolist` commands. 

`tolist()` will convert its entries to a list. But, it can only be called on individual columns (called `Series` in `pandas` documentation)

So, to turn the `NAME` column into a list:

In [42]:
data_table.NAME.tolist()

[u'Lake of the Woods',
 u'Ferry',
 u'Stevens',
 u'Okanogan',
 u'Pend Oreille',
 u'Boundary',
 u'Lincoln',
 u'Flathead',
 u'Glacier',
 u'Toole',
 u'Liberty',
 u'Hill',
 u'Sheridan',
 u'Divide',
 u'Burke',
 u'Renville',
 u'Bottineau',
 u'Rolette',
 u'Towner',
 u'Cavalier',
 u'Pembina',
 u'Kittson',
 u'Roseau',
 u'Blaine',
 u'Phillips',
 u'Valley',
 u'Daniels',
 u'Whatcom',
 u'Bonner',
 u'Ward',
 u'Koochiching',
 u'Skagit',
 u'Williams',
 u'McHenry',
 u'St. Louis',
 u'Roosevelt',
 u'Mountrial',
 u'Marshall',
 u'Ramsey',
 u'Walsh',
 u'Beltrami',
 u'Pierce',
 u'Chelan',
 u'Pondera',
 u'Clallam',
 u'Benson',
 u'Chouteau',
 u'Snohomish',
 u'Island',
 u'Sanders',
 u'Lake',
 u'Nelson',
 u'Grand Forks',
 u'Polk',
 u'Pennington',
 u'Douglas',
 u'McKenzie',
 u'Jefferson',
 u'Richland',
 u'Teton',
 u'McCone',
 u'Shoshone',
 u'Spokane',
 u'Lake',
 u'Clearwater',
 u'Kootenai',
 u'Garfield',
 u'Red Lake',
 u'Grant',
 u'Lincoln',
 u'Lewis and Clark',
 u'Kitsap',
 u'Itasca',
 u'Sheridan',
 u'Wells',
 u'

To extract many columns, you must select the columns you want and call their `.values` attribute. 

If we were interested in grabbing all of the `HR` variables in the dataframe, we could first select those column names:

In [50]:
HRs = [col for col in data_table.columns if col.startswith('HR')]

We can use this to focus only on the columns we want:

In [44]:
data_table[HRs]

HR60       HR70       HR80       HR90
0      0.000000   0.000000   8.855827   0.000000
1      0.000000   0.000000  17.208742  15.885624
2      1.863863   1.915158   3.450775   6.462453
3      2.612330   1.288643   3.263814   6.996502
4      0.000000   0.000000   7.770008   7.478033
5      0.000000   0.000000   4.573101   4.000640
6      7.976390   5.536179   5.633168   5.720497
7      1.011173   1.689475   4.490115   2.814460
8     11.529039   9.273857  28.227324   5.500096
9      0.000000   5.708740   0.000000   6.605892
10     0.000000   0.000000   0.000000   0.000000
11     3.574045   3.840688   7.413585   1.888146
12     0.000000   0.000000   0.000000   0.000000
13     0.000000   0.000000   0.000000   0.000000
14     0.000000   0.000000   0.000000   0.000000
15     0.000000   0.000000   0.000000   0.000000
16     2.945942   0.000000   0.000000   0.000000
17     0.000000   0.000000  19.161808   5.219752
18     0.000000   0.000000   0.000000   0.000000
19     0.000000   8.117213   0.000000   0.000000
20     0.000000   0.000000   0.000000   0.000000
21     0.000000   4.864050   0.000000   0.000000
22     0.000000   0.000000   0.000000   2.218377
23     4.119804   9.910312  14.287755  14.863258
24     5.530668   0.000000  12.421589   0.000000
25     5.854801   5.811757   6.504065   4.045798
26     0.000000  10.811980   0.000000   0.000000
27     1.422131   2.439530   4.061193   2.086920
28     2.138534   2.142245   5.518079   3.756292
29     0.708135   1.138434   0.570854   1.150993
...         ...        ...        ...        ...
3055   0.000000   0.000000   2.107526   0.739919
3056   0.000000   0.970572   4.400324   0.825491
3057   0.611430   2.568301   1.974919   2.828994
3058   2.022286   2.033140   0.000000   3.987956
3059   0.850723   1.981012   2.782690   2.260130
3060   1.790825   3.732470   5.757329   4.007173
3061   0.556604   1.060271   1.010560   2.769193
3062   0.000000   1.756836   5.505395   2.907653
3063   0.427592   3.688132   2.625587   0.000000
3064   0.896660   2.704530   4.229303   2.938915
3065   1.051403   5.379304   7.057466   3.424679
3066   2.095434   6.671377   9.243279   7.285916
3067   1.872835   3.951663   5.339935   3.201065
3068   1.917913   6.382639   1.304631   0.000000
3069   1.939789   4.960564   0.000000   6.072530
3070   5.452765  15.156192  24.841012  28.268787
3071   7.520089   9.163383  10.590286   6.443839
3072   7.202448   9.746302  11.850014  12.561604
3073   8.253379  15.655752  21.173432  16.479507
3074   2.181802   3.074760   3.191133   3.300700
3075   4.902862  11.782264   7.680787  18.362582
3076  18.513376  17.133324  15.034136  12.027015
3077   4.159907   4.126434   3.967782   6.585273
3078   5.403098   5.970974   4.127839   2.586787
3079   1.121183   1.096311   2.442074   2.806112
3080   5.046682  13.152054  13.251761   5.521552
3081   3.411368   7.393533  11.453817   8.691999
3082   1.544425   6.023552   5.280349   4.367330
3083   9.302820   1.800148   3.000030   3.727712
3084   3.396162   2.284879   1.194743   2.048855

[3085 rows x 4 columns]

With this, calling `.values` gives an array containing all of the entries in this subset of the table:

In [45]:
data_table[['HR90', 'HR80']].values

array([[  0.        ,   8.85582713],
       [ 15.88562351,  17.20874204],
       [  6.46245315,   3.4507747 ],
       ..., 
       [  4.36732988,   5.2803488 ],
       [  3.72771194,   3.00003   ],
       [  2.04885495,   1.19474313]])

Using the PySAL pdio tools means that if you're comfortable with working in Pandas, you can continue to do so. 

If you're more comfortable using Numpy or raw Python to do your data processing, PySAL's IO tools naturally support this. 

In [136]:

from numpy import sin,cos,arctan2,sqrt,pi # import from numpy

def gcd(loc1, loc2, R=3961):
    """Great circle distance
    
    Parameters
    ----------
    
    loc1: tuple (long, lat in decimal degrees)
    
    loc2: tuple (long, lat in decimal degrees)
    
    R: Radius of the earth (3961 miles, 6367 km)
    
    Returns
    -------
    great circle distance between loc1 and loc2 in units of R
    
    
    Notes
    ------
    Does not take into account non-spheroidal shape of the Earth
    
    Source: http://andrew.hedges.name/experiments/haversine/
    
    >>> san_diego = -117.1611, 32.7157
    >>> austin = -97.7431, 30.2672
    >>> gcd(san_diego, austin)
    1155.4746441646942
    
    """
    lon1, lat1 = loc1
    lon2, lat2 = loc2
    lon1 = lon1 * pi / 180.0
    lon2 = lon2 * pi / 180.0
    lat1 = lat1 * pi / 180.0
    lat2 = lat2 * pi / 180.0
    dlat = lat2 - lat1
    dlon = lon2 - lon1
    a = (sin(dlat/2.))**2 + cos(lat1) * cos(lat2) * (sin(dlon/2.))**2 
    c = 2 * arctan2( sqrt(a), sqrt(1-a) )
    return R * c
    

In [139]:
san_diego = -117.1611, 32.7157
austin = -97.7431, 30.2672
gcd(san_diego, austin)

1155.4746441646942

In [141]:

import scipy.spatial.distance as d
def near_target_point(polygon, target=austin, threshold=10):
    return gcd(polygon.centroid, target) < threshold 

data_table[data_table.geometry.apply(near_target_point)]

NAME STATE_NAME STATE_FIPS CNTY_FIPS   FIPS  STFIPS  COFIPS  FIPSNO  \
2742  Travis      Texas         48       453  48453      48     453   48453   

      SOUTH      HR60                        ...                          \
2742      1  4.242561                        ...                           

          BLK90      GI59      GI69      GI79      GI89       FH60  FH70  \
2742  10.959791  0.299292  0.372293  0.378953  0.388149  12.976379  10.9   

           FH80       FH90                                           geometry  
2742  14.459691  17.307113  <pysal.cg.shapes.Polygon object at 0x7f85bb32f...  

[1 rows x 70 columns]

## Exercises

1. Find the county with the western most centroid that is withing 1000 miles of Austin.
2. Find the distance between Austin and that centroid.

In [241]:

import scipy.spatial.distance as d
def near_target_point(polygon, target=austin, threshold=1000):
    return gcd(polygon.centroid, target) < threshold 

w1000=data_table[data_table.geometry.apply(near_target_point)]

In [242]:
len(w1000)

1883

In [243]:
g = w1000.geometry

In [244]:
type(g)

pandas.core.series.Series

In [245]:
len(g)

1883

In [246]:
gc = g.copy()


In [247]:
coords = np.array([ p.centroid for p in gc.values])

In [248]:
coords.min(axis=0)

array([-113.93953516,   26.13240654])

In [249]:
west_lon  = np.array([ p.centroid for p in gc.values]).min(axis=0)[0]

In [250]:
west_lon

-113.93953516273253

In [251]:
w1000[w1000.geometry.apply(lambda x: x.centroid[0] == west_lon)]

NAME STATE_NAME STATE_FIPS CNTY_FIPS   FIPS  STFIPS  COFIPS  FIPSNO  \
3080  La Paz    Arizona         04       012  04012       4      12    4012   

      SOUTH      HR60                        ...                          \
3080      0  5.046682                        ...                           

         BLK90      GI59     GI69      GI79      GI89      FH60  FH70  \
3080  2.628811  0.271556  0.36411  0.372662  0.405743  9.216414   8.0   

          FH80       FH90                                           geometry  
3080  9.296093  12.379134  <pysal.cg.shapes.Polygon object at 0x7f85bb246...  

[1 rows x 70 columns]

In [252]:
lapaz = w1000[w1000.NAME=="La Paz"].geometry.values[0].centroid

In [253]:
gcd(lapaz, austin)

972.59627350385097